### ISSUE TITLE
Mono-Native not working on MacOS Catalina

### ISSUE THREAD

**Issue by xDaKillerx** [mono/mono#14793]
## Steps to Reproduce

1. Just run an app . say Sonarr or Radarr
2. 
3. 

### Current Behavior

Opening the app doesnt work and on opening using terminal , results in this error 
```
exception inside UnhandledException handler: (null) assembly:/Library/Frameworks/Mono.framework/Versions/6.5.0/lib/mono/4.5/mscorlib.dll type:TypeInitializationException member:(null)

[ERROR] FATAL UNHANDLED EXCEPTION: System.TypeInitializationException: The type initializer for 'System.Console' threw an exception. ---> System.TypeInitializationException: The type initializer for 'System.ConsoleDriver' threw an exception. ---> System.DllNotFoundException: /Library/Frameworks/Mono.framework/Versions/6.5.0/lib/../lib/libmono-native-compat.dylib assembly:<unknown assembly> type:<unknown type> member:(null)
  at (wrapper managed-to-native) Interop+Sys.Stat(byte&,Interop/Sys/FileStatus&)
  at Interop+Sys.Stat (System.ReadOnlySpan`1[T] path, Interop+Sys+FileStatus& output) [0x0001b] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/external/corefx/src/Common/src/Interop/Unix/System.Native/Interop.Stat.Span.cs:24 
  at System.IO.FileSystem.FileExists (System.ReadOnlySpan`1[T] fullPath, System.Int32 fileType, Interop+ErrorInfo& errorInfo) [0x00007] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/external/corefx/src/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs:445 
  at System.IO.FileSystem.DirectoryExists (System.ReadOnlySpan`1[T] fullPath, Interop+ErrorInfo& errorInfo) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/external/corefx/src/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs:419 
  at System.IO.FileSystem.DirectoryExists (System.ReadOnlySpan`1[T] fullPath) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/external/corefx/src/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs:414 
  at System.IO.Directory.Exists (System.String path) [0x00019] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/external/corefx/src/System.IO.FileSystem/src/System/IO/Directory.cs:57 
  at System.TermInfoDriver.SearchTerminfo (System.String term) [0x00044] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/TermInfoDriver.cs:130 
  at System.TermInfoDriver..ctor (System.String term) [0x0004b] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/TermInfoDriver.cs:162 
  at System.ConsoleDriver.CreateTermInfoDriver (System.String term) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/ConsoleDriver.cs:73 
  at System.ConsoleDriver..cctor () [0x0004d] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/ConsoleDriver.cs:57 
   --- End of inner exception stack trace ---
  at System.Console.SetupStreams (System.Text.Encoding inputEncoding, System.Text.Encoding outputEncoding) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/Console.cs:137 
  at System.Console..cctor () [0x0007d] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/Console.cs:131 
   --- End of inner exception stack trace ---
  at NzbDrone.Common.Instrumentation.GlobalExceptionHandlers.HandleAppDomainException (System.Object sender, System.UnhandledExceptionEventArgs e) [0x0007a] in M:\BuildAgent\work\63739567f01dbcc2\src\NzbDrone.Common\Instrumentation\GlobalExceptionHandlers.cs:48 
```
### Expected Behavior

 successful run

## On which platforms did you notice this

[X] macOS
[ ] Linux
[ ] Windows

**Version Used**:

 6 and even on the latest 6.5

## Stacktrace

```
Please paste the stack trace here if available.
```

hey guys. sorry if i did this wrong . its my first time sharing here.

**Comment by EgorBo** [comment]
Logs for:
```
csc Program.cs
```
https://gist.github.com/EgorBo/042f403ee020375cfc5e208e7ed63dec

**Comment by lambdageek** [comment]
Seems like they removed a symbol `___gss_c_nt_hostbased_service_oid_desc` that libmono-native-compat.dylib depended on:

```
Mono: DllImport error loading library '/Library/Frameworks/Mono.framework/Versions/6.5.0/lib/../lib/libmono-native-compat.dylib': 'dlopen(/Library/Frameworks/Mono.framework/Versions/6.5.0/lib/../lib/libmono-native-compat.dylib, 9): Symbol not found: ___gss_c_nt_hostbased_service_oid_desc
```

**Comment by EgorBo** [comment]
So the problem is clearly somewhere in GSS.framework. I suspect `gss_import_name`

**Comment by filipnavara** [comment]
Yep, it should be related to `GSS_C_NT_HOSTBASED_SERVICE` as used by `gss_import_name`. The latest Heimdal source on Apple Open Source web site still has it but unfortunately they didn't publish source for 10.15 yet.

**Comment by EgorBo** [comment]
The following symbols are missing:
```
gss_OID_desc __gss_c_nt_user_name_oid_desc;
gss_OID_desc __gss_krb5_mechanism_oid_desc;
gss_OID_desc __gss_ntlm_mechanism_oid_desc;
gss_OID_desc __gss_spnego_mechanism_oid_desc;
```
(if I add this to pal_gssapi.c -- it will work).

@filipnavara `nm` returns the same output for me on macos 10.15 + xcode11

**Comment by filipnavara** [comment]
@EgorBo if `nm` shows the symbol then it almost sounds as if it's trying to link against incorrect GSS library.

**Comment by filipnavara** [comment]
I was not able to fully track it down, but here is couple of observations:
- Mono 5.18.1 is not affected since it didn't include the GSS code yet. 
- `otool -L /Library/Frameworks/Mono.framework/Versions/6.5.0/lib/../lib/libmono-native-compat.dylib` does NOT list a reference to the GSS framework even though it should.
- On Mojave the GSS framework got loaded by some of the indirect dependencies like `libcups.2.dylib` which is itself an indirect dependecny of some other library. Later the load of `libmono-native-compat.dylib` succeeds because it can find the symbols in the libraries that were already loaded.
- Local builds on macOS Catalina are messed up on my machine. `./configure` finds the GSS libraries (`HAVE_GSSFW_HEADERS` and `HAVE_GSS_SPNEGO_MECHANISM` are defined in `config.h`) but `ENABLE_GSS` doesn't get set to the correct value. (zsh problem with some syntax maybe?)
- The code for adding the GSS framework reference is https://github.com/mono/mono/blob/a3de0304a190c54124df3486ce9c34b6262787a8/configure.ac#L6193-L6195. I am unable to test it due to the issue above.

**Comment by xDaKillerx** [comment]
@filipnavara yeah confirmed that 5.18.1 works fine with Catalina and the apps ... unfortunately I don't understand the rest :P . Sorry if I can't be more of help ...

**Comment by baulig** [comment]
Thanks a lot for investigating this.

I believe we're using the header files from `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/GSS.framework` - is that correct.

Yesterday, I disassembled the `libmono-native-compat.a` and the symbol comes from `_NetSecurityNative_ImportPrincipalName`:

```
_NetSecurityNative_ImportPrincipalName:
     300:       55      pushq   %rbp
     301:       48 89 e5        movq    %rsp, %rbp
     304:       41 57   pushq   %r15
     306:       41 56   pushq   %r14
     308:       41 54   pushq   %r12
     30a:       53      pushq   %rbx
     30b:       48 83 ec 10     subq    $16, %rsp
     30f:       49 89 cf        movq    %rcx, %r15
     312:       41 89 d4        movl    %edx, %r12d
     315:       48 89 f3        movq    %rsi, %rbx
     318:       49 89 fe        movq    %rdi, %r14
     31b:       4d 85 f6        testq   %r14, %r14
     31e:       74 5e   je      94 <_NetSecurityNative_ImportPrincipalName+0x7e>
     320:       48 85 db        testq   %rbx, %rbx
     323:       74 78   je      120 <_NetSecurityNative_ImportPrincipalName+0x9d>
     325:       4d 85 ff        testq   %r15, %r15
     328:       0f 84 8e 00 00 00       je      142 <_NetSecurityNative_ImportPrincipalName+0xbc>
     32e:       49 83 3f 00     cmpq    $0, (%r15)
     332:       0f 85 a3 00 00 00       jne     163 <_NetSecurityNative_ImportPrincipalName+0xdb>
     338:       be 2f 00 00 00  movl    $47, %esi
     33d:       48 89 df        movq    %rbx, %rdi
     340:       e8 00 00 00 00  callq   0 <_NetSecurityNative_ImportPrincipalName+0x45>
                0000000000000341:  X86_64_RELOC_BRANCH  _strchr
     345:       48 85 c0        testq   %rax, %rax
     348:       48 8b 15 00 00 00 00    movq    (%rip), %rdx
                000000000000034b:  X86_64_RELOC_GOT_LOAD        ___gss_krb5_nt_principal_name_oid_desc@GOTPCREL
     34f:       48 0f 44 15 00 00 00 00         cmoveq  (%rip), %rdx
                0000000000000353:  X86_64_RELOC_GOT     ___gss_c_nt_hostbased_service_oid_desc@GOTPCREL
     357:       44 89 e0        movl    %r12d, %eax
     35a:       48 89 45 d0     movq    %rax, -48(%rbp)
     35e:       48 89 5d d8     movq    %rbx, -40(%rbp)
     362:       48 8d 75 d0     leaq    -48(%rbp), %rsi
     366:       4c 89 f7        movq    %r14, %rdi
     369:       4c 89 f9        movq    %r15, %rcx
     36c:       e8 00 00 00 00  callq   0 <_NetSecurityNative_ImportPrincipalName+0x71>
                000000000000036d:  X86_64_RELOC_BRANCH  _gss_import_name
     371:       48 83 c4 10     addq    $16, %rsp
     375:       5b      popq    %rbx
     376:       41 5c   popq    %r12
     378:       41 5e   popq    %r14
     37a:       41 5f   popq    %r15
     37c:       5d      popq    %rbp
     37d:       c3      retq
```

The source code for this is in `src/Native/Unix/System.Net.Security.Native/pal_gssapi.c` and it's using the macros

```
    if (strchr(inputName, '/') != NULL)
    {
        nameType = GSS_KRB5_NT_PRINCIPAL_NAME;
    }
    else
    {
        nameType = GSS_C_NT_HOSTBASED_SERVICE;
    }
```

These are defined in `/Applications/Xcode.app/Contents//Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/GSS.framework/Versions/A/Headers/gssapi_oid.h`.

**Comment by baulig** [comment]
I can confirm that `otool -L` does not include `GSS`:

```
$ otool -L /Library/Frameworks/Mono.framework/Versions/Current/lib//libmono-native-compat.0.dylib
/Library/Frameworks/Mono.framework/Versions/Current/lib//libmono-native-compat.0.dylib:
	/Library/Frameworks/Mono.framework/Versions/6.0.0/lib/libmono-native-compat.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.50.2)
```

Wouldn't linking with `-framework GSS` do that?  I'm trying to understand whether

1. We are building the library incorrectly.
2. Apple broke something by removing a symbol that was used by a macro in one of their framework header files.
3. How to provide a safe upgrade path for our users.

**Comment by filipnavara** [comment]
> Wouldn't linking with -framework GSS do that? 

It should.

> We are building the library incorrectly.

Most likely it is built incorrectly. I verified that for other .dylib files otool properly reports the GSS framework reference. Unfortunately I was not able to fix my local builds to find out where the problem comes from.

**Comment by EgorBo** [comment]
@filipnavara @baulig I'll add `-framework GSS` for macOS build (currently in confugure.ac it's defined only for iOS) in https://github.com/mono/mono/pull/14820

**Comment by xDaKillerx** [comment]
oh also. just noticed that it seems like the apps cant see the files i have . like it can see the folders etc but they cant see any files. i think its because of the new MacOS changes that gave specific permissions for apps to access storage etc. tried giving both apps "Full Disk Access" on Privacy tab on System Preferences but no go. i think its because its using Mono to run those apps. i even tried giving the library, exec    full disk access. but still no go, any suggestions? Thanks !!

**Comment by EgorBo** [comment]
@xDaKillerx can you share a repro or more details about that? what folder are you scanning in your mono app and what API you are using?

**Comment by xDaKillerx** [comment]
@EgorBo sorry for the late response. The folder is on Desktop. I don't really know what's my API. I don't really understand these stuff. Sorry. But it is like that on both Sonarr and Radarr . Tried giving mono or the .app (for both apps) full disk permissions and still no go.

**Comment by formerandroider** [comment]
@EgorBo there's some discussion in [this thread](https://forums.sonarr.tv/t/problem-after-upgrading-to-the-new-apple-macos-beta-catalina/22514), though it is unrelated to this report.

**Comment by barracuda156** [comment]
> @filipnavara @baulig I'll add `-framework GSS` for macOS build (currently in confugure.ac it's defined only for iOS) in #14820

@EgorBo Unsurprisingly, unconditionally added flag has broken the build, since GSS framework is only available in 10.14+.

**Comment by EgorBo** [comment]
> > @filipnavara @baulig I'll add `-framework GSS` for macOS build (currently in confugure.ac it's defined only for iOS) in #14820
> 
> @EgorBo Unsurprisingly, unconditionally added flag has broken the build, since GSS framework is only available in 10.14+.

The last version of .NET to support macos 10.14 is .NET Core 3.0

**Comment by barracuda156** [comment]
> The last version of .NET to support macos 10.14 is .NET Core 3.0

Given that mono itself compiles on 10.6 (though not yet `mcs`), I am pretty sure earlier macOS can be supported.
